### ISSUE TITLE
eslint: dependency wrongly mark as unused

### ISSUE THREAD

**Issue by znarf** [depcheck/depcheck#500]
This is likely introduced by https://github.com/depcheck/depcheck/issues/476 and https://github.com/depcheck/depcheck/pull/477

Back in this previous issue, the contributor (@sveyret) wrote:

> when analyzing eslint configuration dependencies, it does not look at this dependencies' dependencies.

It should not, seems to me this is defeating the general behavior of `depcheck`. Any dependency that is explicitly used or imported in the project should be declared in `package.json`, even if it's already made available by another dependency. I just spent some time troubleshooting this in a project.

I had the following eslint config:

```
extends:
  - standard-jsx
  - standard-react
```

And the following lines in `package.json`:

```
    "eslint-config-standard": "^12.0.0",
    "eslint-config-standard-jsx": "^8.1.0",
    "eslint-config-standard-react": "^9.0.0",
```

And depcheck was outputing:

```
Unused devDependencies
* eslint-config-standard-jsx
```

Because `eslint-config-standard-jsx` is a dependency of `eslint-config-standard-react`, it's marked as 'unused'. I don't think that's right.

**Comment by sveyret** [comment]
I agree if dependencies' dependencies are dev dependencies, not if they are direct dependencies. Let me explain my use case : I have a project in which I define the way all my company's projects should behave, using an eslint configuration. In this project, I include (as runtime dependencies) all the modules I need for the rules described in it.
In the dependent projects, I only want to add this project I defined, because it should contain everything needed. I don't want to modify all my dependent projects each time I add a rule which requires a new dependency.

**Comment by znarf** [comment]
I get what you want, but seems to me this is not the way depcheck is working generally. Example: when using `lodash`, it's not because some of my dependencies are using `lodash` that I should not add `lodash` to my package.json.

In your use case, seems to me you should avoid having the sub-dependencies  being detected by depcheck. Let's say your config is `eslint-config-mycompany` and it's using internally `eslint-config-node` and `eslint-config-react`. In your project, you should only configure with `eslint-config-mycompany` and not have `eslint-config-node` and `eslint-config-react` mentioned anywhere. This way, depcheck will not be reporting it.

Would be great if you could point to a sample repository so we could see your use case more clearly.

**Comment by sveyret** [comment]
You're right. I don't remember right now what was not working this way…

I think that the repository I'm talking about is public. You can find it there :

https://git.duniter.org/gecogvidanto/gecogvidanto

**Comment by sveyret** [comment]
Of course, this issue is real anyway, your dependency should not be detected as unused, but the correction is not so simple as it seems.

**Comment by znarf** [comment]
Thank you for sharing your case. If we want to look at a single specific example?

- `git clone https://git.duniter.org/gecogvidanto/gecogvidanto`
- `cd gecogvidanto`
- `depcheck core/shared`

Is that what we should look at?

**Comment by sveyret** [comment]
@znarf not exactly, because it is actually a monorepo and it has a specific depcheck configuration which it is called using a (wrapper) binary. If you want to test, you can do:

    git clone https://git.duniter.org/gecogvidanto/gecogvidanto
    npx lerna bootstrap
    cd packages/core/shared
    npm run test:dep

**Comment by sveyret** [comment]
For information, I'll take the point tomorrow if I can find time.

**Comment by znarf** [comment]
I hit some blockers when trying to reproduce with your repo:
- `npx lerna bootstrap` didn't install dependencies in sub-packages,  I tried `yarn install` instead
- `@gecogvidanto/depcheck` was not initially built, I manually built it with `npm run build`

Then:
- `npm run test:dep` was not outputting any failure

From my side, without a reproduction, I'm still having difficulties understanding the initial issue precisely.

**Comment by sveyret** [comment]
Sorry, I think there is a misunderstanding: `test:dep` will not output any failure because it's been corrected in the latest version of `depcheck`. But this correction is creating the issue you found.

**Comment by znarf** [comment]
@sveyret I was under the impression that the version used is 0.9.1 https://git.duniter.org/gecogvidanto/gecogvidanto/blob/master/packages/devtools/depcheck/package.json#L34, the version without the patch, am I missing something?

**Comment by sveyret** [comment]
@znarf you are right. I must I made the correction somewhere else. I'll have to check that.

**Comment by sveyret** [comment]
Hi @znarf, I made a pull request (#506) which should correct your issue. Would you please have time to check it?

**Comment by sveyret** [comment]
@znarf, the PR is now merged. I will let you close the issue if you think it is corrected…

**Comment by znarf** [comment]
@sveyret Thank you, looks good!
